### PR TITLE
ZCS-11499: Regression - Message written in CP932 is displayed incorrectly (garbled)

### PR DESCRIPTION
**Pull request by wilfredo-alfaro**
### Solution

The issue described in ZCS-11499 was introduced by the ZCS-11096 implementation for External Email Warning (EEW), so that it would prevent for messages from being saved and rendered correctly under certain character encoding scenarios, even when the EEW feature was turned off. As a solution, further changes have been done on the `ZimbraLmtpBackend `class in order to make it compatible with all charsets described in the `MimeConstants `class rather than expecting a default charset, which was the root cause for the issue.

### Notes

This issue was first identified from incompatibility the CP932 charset, also known as Windows-31J, and the solution fixes that. However, it also works for other charsets, including:

- us-ascii
- utf-8
- iso-8859-1
- windows-1252
- euc_cn
- gb2312
- gbk
- Windows-31J
- shift_jis

### Testing Done

Please go to ticket for further details.

### CODE REVIEW

File: store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java [hunk 1]
```diff
@@ -339,11 +347,22 @@ public void deliver(LmtpEnvelope env, InputStream in, int sizeHint) throws Unrec
...
-            inEEW = IOUtils.toInputStream(contentEEW);
+            for (int i = 0; i < CHARSET_NAMES.length; i++) {
+                final String charsetName = CHARSET_NAMES[i];
+                // copy input stream and evaluate content encoding
+                final byte[] ba = IOUtils.toByteArray(in);
+                IOUtils.closeQuietly(in);
+                in = new ByteArrayInputStream(ba);
+                final String content = new String(ba, charsetName);
+                // if current encoding is suitable, use it; otherwise, continue or use default
+                if (content.contains(GARBLED_CHARACTER) && i < CHARSET_NAMES.length - 1) {
```
Comment (shrutig0510) [inline store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java:358]: Why do we need this "i < CHARSET_NAMES.length - 1" check here?

Comment (swaatiTelus) [reply store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java:358]: As per the understanding of logic, for the last element of array i.e. `MimeConstants.P_CHARSET_DEFAULT`, Code has to be always executed in else block, that is the reason `i < CHARSET_NAMES.length - 1` is implemented.

Comment (zimsuchitgupta) [reply store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java:358]: We may write with single if block .
`if (!content.contains(GARBLED_CHARACTER) || i == CHARSET_NAMES.length - 1) {
}`

**Review by preshitaagrawal — APPROVED** [review]

File: store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java (hunk 1 above)
Comment (gabhinavTI) [reply store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java]: @zimsuchitgupta @shrutig0510 modified as suggested

File: store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java [hunk 2]
```diff
@@ -339,11 +347,21 @@ public void deliver(LmtpEnvelope env, InputStream in, int sizeHint) throws Unrec
...
-            inEEW = IOUtils.toInputStream(contentEEW);
+            for (int i = 0; i < CHARSET_NAMES.length; i++) {
+                final String charsetName = CHARSET_NAMES[i];
+                // copy input stream and evaluate content encoding
+                final byte[] ba = IOUtils.toByteArray(in);
+                IOUtils.closeQuietly(in);
+                in = new ByteArrayInputStream(ba);
+                final String content = new String(ba, charsetName);
+                // if current encoding is suitable, use it; otherwise, continue or use default
+                if (!content.contains(GARBLED_CHARACTER) || i == CHARSET_NAMES.length - 1) { }
```
Comment (zimsuchitgupta) [inline store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java:358]: I doesn't mean this. You should write this only.
`if (!content.contains(GARBLED_CHARACTER) || i == CHARSET_NAMES.length - 1) { 
       final String contentEEW = ExternalEmailWarning.getInstance().getUpdatedContent(content);
       inEEW = IOUtils.toInputStream(contentEEW, charsetName);
       break;
}`

**Review by zimsuchitgupta — CHANGES_REQUESTED** [review]

File: store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java (hunk 2 above)
Comment (shrutig0510) [reply store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java:358]: Yes....expecting this without else bock

Comment (RakeshAMore) [inline store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java:358]: why is this empty block?
do you mean if the variable content contains `GARBLED_CHARACTER ` and `i` value lesser than CHARSET names? then you want to execute the below code and break the loop?
```
final String contentEEW = ExternalEmailWarning.getInstance().getUpdatedContent(content);
inEEW = IOUtils.toInputStream(contentEEW, charsetName);
break;
```
this seems to be very confusing to me.

Comment (gabhinavTI) [reply store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java]: @zimsuchitgupta @shrutig0510 , Could you please check now

**Review by zimsuchitgupta — APPROVED** [review]

File: store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java (hunk 2 above)
Comment (gabhinavTI) [reply store/src/java/com/zimbra/cs/lmtpserver/ZimbraLmtpBackend.java]: @RakeshAMore , Updated as per Suchit's comment

**Review by shrutig0510 — APPROVED** [review]

**Review by deepakzimb — APPROVED** [review]

**Pull request merged**